light: fix txpool causes panic on reorgOnNewHead

meowsbits opened the pull request:
Fixes #53 

```
panic: runtime error: invalid memory address or nil pointer dereference
Mar 04 12:22:51 sf-etclabs sh[26575]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x1c0 pc=0xc5937c]
Mar 04 12:22:51 sf-etclabs sh[26575]: goroutine 57 [running]:
Mar 04 12:22:51 sf-etclabs sh[26575]: github.com/ethereum/go-ethereum/light.(*TxPool).reorgOnNewHead(0xc00038cc60, 0x11a2960, 0xc0204bd920, 0xc0209ec480,
0xc0204bd920, 0xc02096eb00, 0x8)
Mar 04 12:22:51 sf-etclabs sh[26575]:         /root/go/src/github.com/ethereum/go-ethereum/light/txpool.go:230 +0x23c
Mar 04 12:22:51 sf-etclabs sh[26575]: github.com/ethereum/go-ethereum/light.(*TxPool).setNewHead(0xc00038cc60, 0xc0209ec480)
Mar 04 12:22:51 sf-etclabs sh[26575]:         /root/go/src/github.com/ethereum/go-ethereum/light/txpool.go:311 +0x13a
Mar 04 12:22:51 sf-etclabs sh[26575]: github.com/ethereum/go-ethereum/light.(*TxPool).eventLoop(0xc00038cc60)
Mar 04 12:22:51 sf-etclabs sh[26575]:         /root/go/src/github.com/ethereum/go-ethereum/light/txpool.go:292 +0x53
Mar 04 12:22:51 sf-etclabs sh[26575]: created by github.com/ethereum/go-ethereum/light.NewTxPool
Mar 04 12:22:51 sf-etclabs sh[26575]:         /root/go/src/github.com/ethereum/go-ethereum/light/txpool.go:109 +0x3eb
```

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits commented on .github/FUNDING.yml:
```diff
@@ -0,0 +1 @@
+github: sorpaas
```
Fixed w https://github.com/etclabscore/multi-geth/pull/54/commits/49faa47dc2cd1f6526b40576ba965e6963cedef7

meowsbits committed:
light: fix txpool causes panic on reorgOnNewHead

panic: runtime error: invalid memory address or nil pointer dereference
Mar 04 12:22:51 sf-etclabs sh[26575]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x1c0 pc=0xc5937c]
Mar 04 12:22:51 sf-etclabs sh[26575]: goroutine 57 [running]:
Mar 04 12:22:51 sf-etclabs sh[26575]: github.com/ethereum/go-ethereum/light.(*TxPool).reorgOnNewHead(0xc00038cc60, 0x11a2960, 0xc0204bd920, 0xc0209ec480,
0xc0204bd920, 0xc02096eb00, 0x8)
Mar 04 12:22:51 sf-etclabs sh[26575]:         /root/go/src/github.com/ethereum/go-ethereum/light/txpool.go:230 +0x23c
Mar 04 12:22:51 sf-etclabs sh[26575]: github.com/ethereum/go-ethereum/light.(*TxPool).setNewHead(0xc00038cc60, 0xc0209ec480)
Mar 04 12:22:51 sf-etclabs sh[26575]:         /root/go/src/github.com/ethereum/go-ethereum/light/txpool.go:311 +0x13a
Mar 04 12:22:51 sf-etclabs sh[26575]: github.com/ethereum/go-ethereum/light.(*TxPool).eventLoop(0xc00038cc60)
Mar 04 12:22:51 sf-etclabs sh[26575]:         /root/go/src/github.com/ethereum/go-ethereum/light/txpool.go:292 +0x53
Mar 04 12:22:51 sf-etclabs sh[26575]: created by github.com/ethereum/go-ethereum/light.NewTxPool
Mar 04 12:22:51 sf-etclabs sh[26575]:         /root/go/src/github.com/ethereum/go-ethereum/light/txpool.go:109 +0x3eb

Signed-off-by: meows <b5c6@protonmail.com>

tzdybal commented on light/txpool.go:
```diff
@@ -224,30 +224,38 @@ func (pool *TxPool) reorgOnNewHead(ctx context.Context, newHeader *types.Header)
-	for _, hash := range oldHashes {
-		pool.rollbackTxs(hash, txc)
+
+	if newHeader == nil {
```
This statement doesn't make much sense, especially in here.
Above, we assign `newHeader` to `newh`, and operate on `newh`. If `newHeader` was indeed `nil`, this would fail (except degenerated case when both `newh`/`newHeader` and `oldHeader` are `nil`).

If it actually can happen, or we want to be extra defensive, move the check above line 229.

meowsbits replied:
I was assuming `oldh` was nil...

meowsbits replied:
PTAL at changes thru https://github.com/etclabscore/core-geth/pull/54/commits/687e62601410c098332757dec8291df50daa4ffe.

Have added some test cases and simplified the checks.

tzdybal reviewed (changes requested):
Is it possible to easily reproduce the issue in the test?

meowsbits committed:
light: revert f6ed5d4bbc55afd2918b7847231f8a3f126d1fbd

The check for newHeader nilness should happen earlier,
somewhere up the Chain Insertion line.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
core: add headerchain checks for nil

In real life, I think the ValidateHeaderChain
check is sufficient, but tests use the InsertChain method
without the Validate_ method prior, so that's why
I think both these checks are needed.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
light: add txpool edge case tests and fix reorg method can panic

Adds a few edge-cases tests in TestTxPool which test:
- non-nil txpool header-hash (by provenance) returns a nil value from chaindb
  This could be from a bad chaindb or an otherwise fucked pool head hash.

  + The otherwise fucked case is also tested.
- pool attempts to set head to a nil value

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
light: handle case if chain is at or near geneesis

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
light: remove typo dupe line, use Fatal instead of panic

Signed-off-by: meows <b5c6@protonmail.com>
